File: packages/unit-conversion/dotnet/NavArch.UnitConversion/Services/UnitConverter.cs
using System.Globalization;
using NavArch.UnitConversion.Models;
using NavArch.UnitConversion.Providers;

namespace NavArch.UnitConversion.Services;

/// <summary>
/// Main unit conversion service
/// </summary>
public class UnitConverter : IUnitConverter
{
    private readonly XmlUnitSystemProvider _provider;
    private readonly Dictionary<string, UnitSystemDefinition> _systems;
    private readonly Dictionary<(string, string), decimal> _conversionMatrix;

    public UnitConverter(string? xmlConfigPath = null)
    {
        _provider = new XmlUnitSystemProvider(xmlConfigPath);
        _systems = _provider.LoadUnitSystems();
        _conversionMatrix = _provider.LoadConversionMatrix();
    }

    public decimal Convert(decimal value, string fromSystem, string toSystem, string category)
    {
        if (fromSystem == toSystem)
            return value;

        // Get unit IDs for the category in each system
        var fromUnit = GetUnitIdForCategory(fromSystem, category);
        var toUnit = GetUnitIdForCategory(toSystem, category);

        if (fromUnit == null || toUnit == null)
            throw new ArgumentException($"Category '{category}' not found in one of the unit systems");

        // Look up conversion factor
        if (_conversionMatrix.TryGetValue((fromUnit, toUnit), out var factor))
        {
            return value * factor;
        }

        throw new InvalidOperationException($"No conversion available from {fromUnit} to {toUnit}");
    }

    public Dictionary<string, decimal> ConvertBatch(
        Dictionary<string, (decimal value, string category)> values,
        string fromSystem,
        string toSystem)
    {
        var results = new Dictionary<string, decimal>();

        foreach (var (key, (value, category)) in values)
        {
            results[key] = Convert(value, fromSystem, toSystem, category);
        }

        return results;
    }

    public string GetUnitSymbol(string unitSystem, string category, string locale = "en")
    {
        var unit = GetUnitDefinition(unitSystem, category);
        return unit?.Symbol ?? string.Empty;
    }

    public string GetUnitName(string unitSystem, string category, string locale = "en", bool plural = false)
    {
        var unit = GetUnitDefinition(unitSystem, category);
        if (unit == null)
            return string.Empty;

        var names = plural ? unit.PluralNames : unit.Names;

        // Try to get localized name, fall back to English
        if (names.TryGetValue(locale, out var name))
            return name;

        if (names.TryGetValue("en", out var englishName))
            return englishName;

        return unit.Id;
    }

    public string GetCategoryName(string category, string locale = "en")
    {
        // Try to find the category in any system (they should have same names)
        foreach (var system in _systems.Values)
        {
            if (system.Categories.TryGetValue(category, out var categoryDef))
            {
                if (categoryDef.Names.TryGetValue(locale, out var name))
                    return name;

                if (categoryDef.Names.TryGetValue("en", out var englishName))
                    return englishName;

                return category;
            }
        }

        return category;
    }

    public List<UnitSystemInfo> GetAvailableUnitSystems(string locale = "en")
    {
        return _systems.Values.Select(system => GetUnitSystemInfo(system.Id, locale)).ToList();
    }

    public UnitSystemInfo GetUnitSystemInfo(string unitSystemId, string locale = "en")
    {
        if (!_systems.TryGetValue(unitSystemId, out var system))
            throw new ArgumentException($"Unit system '{unitSystemId}' not found");

        var name = system.Names.TryGetValue(locale, out var localizedName)
            ? localizedName
            : system.Names.GetValueOrDefault("en", unitSystemId);

        var description = system.Descriptions.TryGetValue(locale, out var localizedDesc)
            ? localizedDesc
            : system.Descriptions.GetValueOrDefault("en", string.Empty);

        return new UnitSystemInfo(
            unitSystemId,
            name,
            description,
            system.IsDefault,
            system.Categories.Keys.ToList());
    }

    public string FormatValue(decimal value, string unitSystem, string category, string locale = "en", int decimals = 2)
    {
        var culture = GetCultureInfo(locale);
        var formattedValue = value.ToString($"N{decimals}", culture);
        var symbol = GetUnitSymbol(unitSystem, category, locale);

        return $"{formattedValue} {symbol}";
    }

    private string? GetUnitIdForCategory(string unitSystem, string category)
    {
        if (!_systems.TryGetValue(unitSystem, out var system))
            return null;

        if (!system.Categories.TryGetValue(category, out var categoryDef))
            return null;

        // Return the first (and typically only) unit in this category
        return categoryDef.Units.FirstOrDefault()?.Id;
    }

    private UnitDefinition? GetUnitDefinition(string unitSystem, string category)
    {
        if (!_systems.TryGetValue(unitSystem, out var system))
            return null;

        if (!system.Categories.TryGetValue(category, out var categoryDef))
            return null;

        return categoryDef.Units.FirstOrDefault();
    }

    private static CultureInfo GetCultureInfo(string locale)
    {
        try
        {
            return CultureInfo.GetCultureInfo(locale);
        }
        catch
        {
            return CultureInfo.InvariantCulture;
        }
    }
}

File: packages/unit-conversion/typescript/src/UnitConverter.ts
import type {
  UnitSystemId,
  UnitCategory,
  Locale,
  UnitSystemInfo,
  UnitInfo,
} from './models';
import { UNIT_SYSTEMS, CONVERSION_MATRIX } from './config';

/**
 * Main unit conversion service
 */
export class UnitConverter {
  /**
   * Convert a value between unit systems for a specific category
   */
  convert(
    value: number,
    fromSystem: UnitSystemId,
    toSystem: UnitSystemId,
    category: UnitCategory
  ): number {
    if (fromSystem === toSystem) {
      return value;
    }

    const fromUnitId = this.getUnitIdForCategory(fromSystem, category);
    const toUnitId = this.getUnitIdForCategory(toSystem, category);

    if (!fromUnitId || !toUnitId) {
      throw new Error(`Category '${category}' not found in one of the unit systems`);
    }

    const factor = CONVERSION_MATRIX[fromUnitId]?.[toUnitId];
    if (factor === undefined) {
      throw new Error(`No conversion available from ${fromUnitId} to ${toUnitId}`);
    }

    return value * factor;
  }

  /**
   * Convert multiple values in batch
   */
  convertBatch(
    values: Record<string, { value: number; category: UnitCategory }>,
    fromSystem: UnitSystemId,
    toSystem: UnitSystemId
  ): Record<string, number> {
    const results: Record<string, number> = {};

    for (const [key, { value, category }] of Object.entries(values)) {
      results[key] = this.convert(value, fromSystem, toSystem, category);
    }

    return results;
  }

  /**
   * Get unit symbol for a category in a specific unit system
   */
  getUnitSymbol(unitSystem: UnitSystemId, category: UnitCategory, locale: Locale = 'en'): string {
    const unit = this.getUnitDefinition(unitSystem, category);
    return unit?.symbol ?? '';
  }

  /**
   * Get unit name for a category in a specific unit system
   */
  getUnitName(
    unitSystem: UnitSystemId,
    category: UnitCategory,
    locale: Locale = 'en',
    plural: boolean = false
  ): string {
    const unit = this.getUnitDefinition(unitSystem, category);
    if (!unit) return '';

    const names = plural ? unit.pluralNames : unit.names;
    return names[locale] ?? names.en ?? unit.id;
  }

  /**
   * Get category name
   */
  getCategoryName(category: UnitCategory, locale: Locale = 'en'): string {
    // Get from SI system (they're the same across systems)
    const categoryDef = UNIT_SYSTEMS.SI.categories[category];
    return categoryDef?.names[locale] ?? categoryDef?.names.en ?? category;
  }

  /**
   * Get all available unit systems
   */
  getAvailableUnitSystems(locale: Locale = 'en'): UnitSystemInfo[] {
    return Object.values(UNIT_SYSTEMS).map((system) =>
      this.getUnitSystemInfo(system.id, locale)
    );
  }

  /**
   * Get unit system information
   */
  getUnitSystemInfo(unitSystemId: UnitSystemId, locale: Locale = 'en'): UnitSystemInfo {
    const system = UNIT_SYSTEMS[unitSystemId];
    if (!system) {
      throw new Error(`Unit system '${unitSystemId}' not found`);
    }

    return {
      id: system.id,
      name: system.names[locale] ?? system.names.en ?? system.id,
      description: system.descriptions[locale] ?? system.descriptions.en ?? '',
      isDefault: system.isDefault,
      categories: Object.keys(system.categories) as UnitCategory[],
    };
  }

  /**
   * Format a value with localized unit
   */
  formatValue(
    value: number,
    unitSystem: UnitSystemId,
    category: UnitCategory,
    locale: Locale = 'en',
    decimals: number = 2
  ): string {
    const formattedValue = this.formatNumber(value, locale, decimals);
    const symbol = this.getUnitSymbol(unitSystem, category, locale);

    return `${formattedValue} ${symbol}`;
  }

  /**
   * Format number according to locale
   */
  private formatNumber(value: number, locale: Locale, decimals: number): string {
    return value.toLocaleString(locale === 'es' ? 'es-ES' : 'en-US', {
      minimumFractionDigits: decimals,
      maximumFractionDigits: decimals,
    });
  }

  /**
   * Get unit ID for a category in a unit system
   */
  private getUnitIdForCategory(unitSystem: UnitSystemId, category: UnitCategory): string | null {
    const system = UNIT_SYSTEMS[unitSystem];
    if (!system) return null;

    const categoryDef = system.categories[category];
    if (!categoryDef) return null;

    return categoryDef.units[0]?.id ?? null;
  }

  /**
   * Get unit definition
   */
  private getUnitDefinition(unitSystem: UnitSystemId, category: UnitCategory) {
    const system = UNIT_SYSTEMS[unitSystem];
    if (!system) return null;

    const categoryDef = system.categories[category];
    if (!categoryDef) return null;

    return categoryDef.units[0] ?? null;
  }
}

// Export a singleton instance
export const unitConverter = new UnitConverter();

